src/util/dom.js

```diff
@@ -53,19 +53,19 @@ exports.contains = contains
 
 let accumulatedCSS = "", cssNode = null
 
-function insertCSS(css) {
-  if (cssNode) cssNode.textContent += css
+function insertCSS(pm, css) {
+  if (pm.cssNode) pm.cssNode.textContent += css
   else accumulatedCSS += css
 }
 exports.insertCSS = insertCSS
 
 // This is called when a ProseMirror instance is created, to ensure
 // the CSS is in the DOM.
-function ensureCSSAdded() {
-  if (!cssNode) {
-    cssNode = document.createElement("style")
-    cssNode.textContent = "/* ProseMirror CSS */\n" + accumulatedCSS
-    document.head.insertBefore(cssNode, document.head.firstChild)
+function ensureCSSAdded(pm) {
+  if (!pm.cssNode) {
```

Comment:
Indeed.

We caught up in person and agreed there was no other guaranteed safe approach in Shadow DOM than to put it inside `pm.wrapper`, once per instance.

In global DOM (i.e. `document`) it's safe-enough to just share a single `<style>` in `document.head`.